File: src/main/java/aurocosh/divinefavor/common/lib/distributed_random/DistributedRandomEntityList.java
package aurocosh.divinefavor.common.lib.distributed_random;

import aurocosh.divinefavor.DivineFavor;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityList;

import java.util.Map;

public class DistributedRandomEntityList extends DistributedRandomList<Class<? extends Entity>> {

    public DistributedRandomEntityList(Map<String, Double> entityTypeMap) {
        for (Map.Entry<String, Double> entry : entityTypeMap.entrySet()) {
            String entityName = entry.getKey();
            Class<? extends Entity> entityClass = EntityList.getClassFromName(entityName);
            if (entityClass != null)
                add(entityClass, entry.getValue());
            else
                DivineFavor.logger.error("Entity type not found: " + entityName);
        }
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/lib/distributed_random/DistributedRandomEntityList.kt
package aurocosh.divinefavor.common.lib.distributed_random

import aurocosh.divinefavor.DivineFavor
import net.minecraft.entity.Entity
import net.minecraft.util.ResourceLocation

class DistributedRandomEntityList(entityTypeMap: Map<String, Double>) : DistributedRandomList<Class<out Entity>>() {

    init {
        for ((entityName, value) in entityTypeMap) {
            val entityClass = getClassFromName(entityName)
            if (entityClass != null)
                add(entityClass, value)
            else
                DivineFavor.logger.error("Entity type not found: $entityName")
        }
    }

    companion object {
        fun getClassFromName(p_192839_0_: String): Class<out Entity>? {
            val entry = net.minecraftforge.fml.common.registry.ForgeRegistries.ENTITIES.getValue(ResourceLocation(p_192839_0_))
            return entry?.entityClass
        }
    }
}
